src/io/windows.zig

```diff
@@ -0,0 +1,1132 @@
+        nanoseconds: u63,
+    ) void {
+        self.submit(
+            context,
+            callback,
+            completion,
+            .timeout,
+            .{ .deadline = self.timer.monotonic() + nanoseconds },
+            struct {
+                fn do_operation(ctx: Completion.Context, op: anytype) TimeoutError!void {
+                    _ = ctx;
+                    _ = op;
+                    return;
+                }
+            },
+        );
+    }
+
+    pub const INVALID_SOCKET = os.windows.ws2_32.INVALID_SOCKET;
+
+    pub fn open_socket(self: *IO, family: u32, sock_type: u32, protocol: u32) !os.socket_t {
+        var flags: os.windows.DWORD = 0;
+        flags |= os.windows.ws2_32.WSA_FLAG_OVERLAPPED;
+        flags |= os.windows.ws2_32.WSA_FLAG_NO_HANDLE_INHERIT;
+
+        const socket = try os.windows.WSASocketW(
+            @bitCast(i32, family),
+            @bitCast(i32, sock_type),
+            @bitCast(i32, protocol),
+            null,
+            0,
+            flags,
+        );
+        errdefer os.closeSocket(socket);
+
+        const socket_iocp = try os.windows.CreateIoCompletionPort(socket, self.iocp, 0, 0);
+        assert(socket_iocp == self.iocp);
+
+        // Ensure that synchronous IO completion doesn't queue an unneeded overlapped
+        // and that the event for the socket (WaitForSingleObject) doesn't need to be set.
+        var mode: os.windows.BYTE = 0;
+        mode |= os.windows.FILE_SKIP_COMPLETION_PORT_ON_SUCCESS;
+        mode |= os.windows.FILE_SKIP_SET_EVENT_ON_HANDLE; 
+
+        const handle = @ptrCast(os.windows.HANDLE, socket);
+        try os.windows.SetFileCompletionNotificationModes(handle, mode);
+
+        return socket;
+    }
+
+    pub fn open_dir(dir_path: [:0]const u8) !os.fd_t {
+        const dir = try std.fs.cwd().openDirZ(dir_path, .{});
+        return dir.fd;
+    }
+
+    pub fn open_file(
+        self: *IO,
+        dir_handle: os.fd_t,
+        relative_path: [:0]const u8,
+        size: u64,
+        must_create: bool,
+    ) !os.fd_t {
+        const path_w = try os.windows.sliceToPrefixedFileW(relative_path);
+
+        // FILE_CREATE = O_CREAT | O_EXCL
+        var creation_disposition: os.windows.DWORD = 0;
+        if (must_create) {
+            log.info("creating \"{s}\"...", .{relative_path});
+            creation_disposition = os.windows.FILE_CREATE;
+        } else {
+            log.info("opening \"{s}\"...", .{relative_path});
+            creation_disposition = os.windows.OPEN_EXISTING;
+        }
+
+        // O_EXCL
+        var shared_mode: os.windows.DWORD = 0;
+
+        // O_RDWR
+        var access_mask: os.windows.DWORD = 0;
+        access_mask |= os.windows.GENERIC_READ;
+        access_mask |= os.windows.GENERIC_WRITE;
+
+        // O_DIRECT
+        var attributes: os.windows.DWORD = 0;
+        attributes |= os.windows.FILE_FLAG_NO_BUFFERING;
+        attributes |= os.windows.FILE_FLAG_WRITE_THROUGH;
```

Comment:
Let's add a comment and assertion after this to explain why `FILE_FLAG_WRITE_THROUGH` is critical:

```zig
        // This is critical as we rely on O_DSYNC for fsync() whenever we write to the file:
        assert((attributes & os.windows.FILE_FLAG_WRITE_THROUGH) > 0);
```